File: src/test/java/com/testainers/LengthResourceTest.java
package com.testainers;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.ContentType;
import io.restassured.http.Method;
import jakarta.ws.rs.core.MediaType;
import org.junit.jupiter.api.Test;

import java.util.List;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

/**
 * @author Eduardo Folly
 */
@QuarkusTest
public class LengthResourceTest {

    static final List<Method> METHODS =
            List.of(Method.GET, Method.POST, Method.PUT,
                    Method.PATCH, Method.DELETE);

    @Test
    public void length0() {
        for (Method method : METHODS) {
            given().when()
                   .request(method, "/length/0")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid size: 0"));
        }
    }

    @Test
    public void length5() {
        for (Method method : METHODS) {
            given().when()
                   .request(method, "/length/5")
                   .then()
                   .statusCode(200)
                   .contentType(ContentType.TEXT)
                   .header("content-length", "5")
                   .body(is("00000"));
        }
    }

    @Test
    public void length10() {
        for (Method method : METHODS) {
            given().when()
                   .accept(MediaType.TEXT_PLAIN)
                   .request(method, "/length/10")
                   .then()
                   .statusCode(200)
                   .contentType(ContentType.TEXT)
                   .header("content-length", "10")
                   .body(is("0000000000"));
        }
    }

    @Test
    public void length1024() {
        for (Method method : METHODS) {
            given().when()
                   .accept(MediaType.APPLICATION_OCTET_STREAM)
                   .request(method, "/length/1024")
                   .then()
                   .statusCode(200)
                   .contentType(ContentType.BINARY)
                   .header("content-length", "1024")
                   .body(notNullValue());
        }
    }

    @Test
    public void length2049() {
        for (Method method : METHODS) {
            given().when()
                   .accept(MediaType.APPLICATION_OCTET_STREAM)
                   .request(method, "/length/2049")
                   .then()
                   .statusCode(500)
                   .contentType(ContentType.TEXT)
                   .body(is("Invalid size: 2049"));
        }
    }

}

File: src/test/kotlin/com/testainers/LengthResourceTest.kt
package com.testainers

import io.quarkus.test.junit.QuarkusTest
import io.restassured.http.ContentType
import io.restassured.http.Method
import io.restassured.module.kotlin.extensions.*
import jakarta.ws.rs.core.MediaType
import org.hamcrest.Matchers.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

/**
 * @author Eduardo Folly
 */
@QuarkusTest
class LengthResourceTest : BaseResourceTest() {
    companion object : BaseResourceTest() {
        @JvmStatic
        fun invalidLength(): List<Arguments> =
            argumentGenerator(listOf(-1, 0, 2049), remove = listOf(Method.HEAD))

        @JvmStatic
        fun successTextLength(): List<Arguments> =
            argumentGenerator(listOf(1, 10), remove = listOf(Method.HEAD))

        @JvmStatic
        fun successOctetLength(): List<Arguments> =
            argumentGenerator(listOf(512, 2048), remove = listOf(Method.HEAD))
    }

    @ParameterizedTest
    @MethodSource("notFoundStatus")
    fun notFound(
        method: Method,
        length: String?,
    ) {
        json(method)
            .request(method, "/length/$length")
            .then()
            .statusCode(404)
            .statusLine(containsStringIgnoringCase("Not Found"))
    }

    @ParameterizedTest
    @MethodSource("invalidLength")
    fun invalid(
        method: Method,
        length: Int,
    ) {
        When {
            request(method, "/length/$length")
        } Then {
            statusCode(500)
            contentType(ContentType.TEXT)
            body(equalTo("Invalid size: $length"))
        }
    }

    @ParameterizedTest
    @MethodSource("successTextLength")
    fun successText(
        it: Method,
        length: Int,
    ) {
        When {
            request(it, "/length/$length")
        } Then {
            statusCode(200)
            contentType(ContentType.TEXT)
            header("Content-Length", "$length")
            body(equalTo("0".repeat(length)))
        }
    }

    @ParameterizedTest
    @MethodSource("successOctetLength")
    fun successOctet(
        it: Method,
        length: Int,
    ) {
        Given {
            given()
            accept(MediaType.APPLICATION_OCTET_STREAM)
        } When {
            request(it, "/length/$length")
        } Then {
            statusCode(200)
            contentType(ContentType.BINARY)
            header("Content-Length", "$length")
            body(notNullValue())
        }
    }
}
